node/coroutines/src/client_task.rs

```diff
@@ -133,9 +133,19 @@ impl Stream for ClientTask {
             }
 
             match self.incoming_block_rx.poll() {
-                Ok(Async::Ready(Some((peer_id, blocks)))) => {
-                    if let ind @ Some(_) = self.try_import_blocks(peer_id, blocks) {
-                        new_block_index = ind;
+                Ok(Async::Ready(Some((peer_id, blocks, best_index)))) => {
+                    if let Some(index) = self.try_import_blocks(peer_id, blocks) {
+                        if index < best_index {
```

Comment:
This code makes certain assumptions. Specifically, it assumes that if we send fetch request for blocks from X to Y then the node will reply with the blocks X...Z, where Z<=Y (otherwise if it replies with T..Z, where X<T<=Z<=Y `index` will not change and we will issue exactly the same fetch request). We are creating a large number of implicit protocol assumptions, that will be hard to manage in a long run if we do not document them. Could we please keep track of all of them as at the minimum sequence diagrams in https://github.com/nearprotocol/nomicon ? Please create MemPool section next to Nightshade and Network, and add subsections with sequence diagrams, similarly to this: https://nearprotocol.github.io/nomicon/Network/Consensus.html (source code: https://raw.githubusercontent.com/nearprotocol/nomicon/master/src/Network/Consensus.md).